server/src/core/user/user-service.ts

```diff
@@ -79,6 +90,35 @@ export class Users implements UserService {
     }
   }
 
+  async nominateBeneficiary(args: UserNominateBeneficiaryArgs): Promise<User> {
+    const { phone, nominator } = args;
+    try {
+      const result = await this.collection.findOneAndUpdate(
+        { phone, roles: { $nin: ['donor'] } }, 
+        { 
+          $addToSet: { roles: 'beneficiary', donors: nominator }, 
+          $currentDate: { updatedAt: true }, 
+          $setOnInsert: { 
+            _id: generateId(), 
+            password: '', 
+            phone, 
+            addedBy: nominator, 
+            createdAt: new Date(),
+          } 
+        },
+        { upsert: true, returnOriginal: false }
+      );
```

Comment:
I think you should add some comments to explain the login behind this query, e.g.:
If the phone number does not yet exist, a new user is created with a beneficiary role and the nominator is added as a donor for that user. If the phone number already exists, then it must not be a donor because a user cannot be both a donor and a beneficiary. If the user is not a donor, then this nominator is added as a donor and the beneficiary role added to that user in case the user was not already a beneficiary.